main/src/util/profiler.hpp

```diff
@@ -0,0 +1,320 @@
+                case CN: efilename = "energy-CN" + std::to_string(numRanks) + "Ranks.txt"; break;
+                default: efilename = "error-energy-file.txt"; break;
+            }
+            energyFile.open(efilename);
+
+            for (int i = 0; i < numRanks; i++)
+            {
+                energyFile << "RANK" << i << ",";
+            }
+            energyFile << std::endl;
+            for (auto& element : vecEnergies)
+            {
+                for (int i = 0; i < numRanks; i++)
+                {
+                    energyFile << element[i] << ",";
+                }
+
+                energyFile << std::endl;
+                delete[] element;
+            }
+
+            std::cout << "Energy data written." << std::endl;
+            energyFile.close();
+        }
+    }
+};
+
+class Profiler
+{
+private:
+    std::vector<float*>        vecFuncTimes; // vector of timesteps for all ranks
+    std::vector<float>         funcTimes;    // vector of timesteps for each rank
+    int                        _numRanks;
+    int                        _rank;
+    H5PartFile*                _profilingFile;
+    H5PartFile*                _energyFile;
+    std::vector<CrayPmtReader> vecEnergyReaders;
+
+public:
+    Profiler(int rank)
+        : _rank(rank)
+    {
+        MPI_Comm_size(MPI_COMM_WORLD, &_numRanks);
+        std::string profilingFilename = "profiling.h5";
+        std::string energyFilename    = "energy.h5";
+
+        if (rank == 0 && std::filesystem::exists(profilingFilename)) { std::filesystem::remove(profilingFilename); }
+        if (rank == 0 && std::filesystem::exists(energyFilename)) { std::filesystem::remove(energyFilename); }
+        MPI_Barrier(MPI_COMM_WORLD);
+
+        _profilingFile = fileutils::openH5Part(profilingFilename, H5PART_WRITE | H5PART_VFD_MPIIO_IND, MPI_COMM_WORLD);
+        _energyFile    = fileutils::openH5Part(energyFilename, H5PART_WRITE | H5PART_VFD_MPIIO_IND, MPI_COMM_WORLD);
+
+        // Create Cray energy readers
+        CrayPmtReader gpuReader(GPU, rank);
+        CrayPmtReader cpuReader(CPU, rank);
+        CrayPmtReader memReader(MEM, rank);
+        CrayPmtReader cnReader(CN, rank);
+        vecEnergyReaders.push_back(std::move(gpuReader));
+        vecEnergyReaders.push_back(std::move(cpuReader));
+        vecEnergyReaders.push_back(std::move(memReader));
+        vecEnergyReaders.push_back(std::move(cnReader));
+    }
+    ~Profiler() {}
+
+    void printProfilingInfoHDF5(size_t timesteps)
+    {
+        H5PartSetStep(_profilingFile, 0);
+        H5PartSetNumParticles(_profilingFile, funcTimes.size());
+        int numFuncs = funcTimes.size() / timesteps;
+        fileutils::writeH5PartStepAttrib(_profilingFile, "NUM_RANKS", &_numRanks, 1);
+        fileutils::writeH5PartStepAttrib(_profilingFile, "NUM_FUNCS", &numFuncs, 1);
+        std::string fieldName = "RANK" + std::to_string(_rank);
+        fileutils::writeH5PartField(_profilingFile, fieldName, funcTimes.data());
```

Comment:
This will give you a random label. For parallel writes, the field name has to match across ranks.

Reply:
I thought HDF5 could write into multiple labels in parallel. Is it not the case?